{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils import Data\n",
    "\n",
    "data_csv = Data(\n",
    "    fGEP_SGA = 'data/CITRUS_GEP_SGAseparated.csv',\n",
    "    fgene_tf_SGA = 'data/CITRUS_gene_tf_SGAseparated.csv',\n",
    "    fcancerType_SGA = 'data/CITRUS_canType_SGAseparated.csv',\n",
    "    fSGA_SGA = 'data/CITRUS_SGA_SGAseparated.csv',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "def shared_tfs(gene_str):\n",
    "    genes = gene_str.split(',')\n",
    "    genes = set([g.rstrip().lstrip() for g in genes])\n",
    "    genes = genes.intersection(set(data_csv.tf))\n",
    "    return genes if len(genes) >0 else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathways2genes = pd.read_csv('/ihome/hosmanbeyoglu/kor11/tools/CITRUS/GenNet/resources/pathways/CPDB_pathways_genes.tab', '\\t')[:-1]\n",
    "kegg_pathways = pd.read_csv('/ihome/hosmanbeyoglu/kor11/tools/CITRUS/GenNet/resources/pathways/pathway_overview_KEGG.csv')\n",
    "pathways2genes = pathways2genes[pathways2genes.source == 'KEGG']\n",
    "pathways2genes.pathway = pathways2genes.pathway.str[:-len('- Homo sapiens (human)')]\n",
    "pathways2genes.pathway = [i.rstrip().lstrip() for i in pathways2genes.pathway]\n",
    "kegg_pathways.local = [i.rstrip().lstrip() for i in kegg_pathways.local]\n",
    "pathways2genes['external_id'] = pathways2genes['external_id'].str[len('path:hsa'):].astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathways2genes['tf'] = pathways2genes['hgnc_symbol_ids'].apply(shared_tfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>global</th>\n",
       "      <th>mid</th>\n",
       "      <th>hsaid</th>\n",
       "      <th>local</th>\n",
       "      <th>mid_id</th>\n",
       "      <th>global_id</th>\n",
       "      <th>local_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>150</td>\n",
       "      <td>Cellular Processes</td>\n",
       "      <td>Cell growth and death</td>\n",
       "      <td>4110</td>\n",
       "      <td>Cell cycle</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>151</td>\n",
       "      <td>Cellular Processes</td>\n",
       "      <td>Cell growth and death</td>\n",
       "      <td>4114</td>\n",
       "      <td>Oocyte meiosis</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>156</td>\n",
       "      <td>Cellular Processes</td>\n",
       "      <td>Cell growth and death</td>\n",
       "      <td>4115</td>\n",
       "      <td>p53 signaling pathway</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>152</td>\n",
       "      <td>Cellular Processes</td>\n",
       "      <td>Cell growth and death</td>\n",
       "      <td>4210</td>\n",
       "      <td>Apoptosis</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>153</td>\n",
       "      <td>Cellular Processes</td>\n",
       "      <td>Cell growth and death</td>\n",
       "      <td>4215</td>\n",
       "      <td>Apoptosis - multiple species</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>229</td>\n",
       "      <td>Organismal Systems</td>\n",
       "      <td>Nervous system</td>\n",
       "      <td>4730</td>\n",
       "      <td>Long-term depression</td>\n",
       "      <td>32</td>\n",
       "      <td>5</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>234</td>\n",
       "      <td>Organismal Systems</td>\n",
       "      <td>Sensory system</td>\n",
       "      <td>4740</td>\n",
       "      <td>Olfactory transduction</td>\n",
       "      <td>36</td>\n",
       "      <td>5</td>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>235</td>\n",
       "      <td>Organismal Systems</td>\n",
       "      <td>Sensory system</td>\n",
       "      <td>4742</td>\n",
       "      <td>Taste transduction</td>\n",
       "      <td>36</td>\n",
       "      <td>5</td>\n",
       "      <td>212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328</th>\n",
       "      <td>233</td>\n",
       "      <td>Organismal Systems</td>\n",
       "      <td>Sensory system</td>\n",
       "      <td>4744</td>\n",
       "      <td>Phototransduction</td>\n",
       "      <td>36</td>\n",
       "      <td>5</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>236</td>\n",
       "      <td>Organismal Systems</td>\n",
       "      <td>Sensory system</td>\n",
       "      <td>4750</td>\n",
       "      <td>Inflammatory mediator regulation of TRP channels</td>\n",
       "      <td>36</td>\n",
       "      <td>5</td>\n",
       "      <td>214</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>330 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0              global                    mid  hsaid  \\\n",
       "0           150  Cellular Processes  Cell growth and death   4110   \n",
       "1           151  Cellular Processes  Cell growth and death   4114   \n",
       "2           156  Cellular Processes  Cell growth and death   4115   \n",
       "3           152  Cellular Processes  Cell growth and death   4210   \n",
       "4           153  Cellular Processes  Cell growth and death   4215   \n",
       "..          ...                 ...                    ...    ...   \n",
       "325         229  Organismal Systems         Nervous system   4730   \n",
       "326         234  Organismal Systems         Sensory system   4740   \n",
       "327         235  Organismal Systems         Sensory system   4742   \n",
       "328         233  Organismal Systems         Sensory system   4744   \n",
       "329         236  Organismal Systems         Sensory system   4750   \n",
       "\n",
       "                                                local  mid_id  global_id  \\\n",
       "0                                          Cell cycle       7          0   \n",
       "1                                      Oocyte meiosis       7          0   \n",
       "2                               p53 signaling pathway       7          0   \n",
       "3                                           Apoptosis       7          0   \n",
       "4                        Apoptosis - multiple species       7          0   \n",
       "..                                                ...     ...        ...   \n",
       "325                              Long-term depression      32          5   \n",
       "326                            Olfactory transduction      36          5   \n",
       "327                                Taste transduction      36          5   \n",
       "328                                 Phototransduction      36          5   \n",
       "329  Inflammatory mediator regulation of TRP channels      36          5   \n",
       "\n",
       "     local_id  \n",
       "0         132  \n",
       "1         133  \n",
       "2         134  \n",
       "3         149  \n",
       "4         152  \n",
       "..        ...  \n",
       "325       210  \n",
       "326       211  \n",
       "327       212  \n",
       "328       213  \n",
       "329       214  \n",
       "\n",
       "[330 rows x 8 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import plotly.graph_objects as go\n",
    "import chart_studio.plotly as py\n",
    "import plotly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 0, 0, 1, 1, 1, 2, 2, 2, 5]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sources[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 2, 3, 4, 5, 6, 5, 6, 4, 7]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "targets[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['install',\n",
       " 'signup',\n",
       " 'purchase',\n",
       " 'reopen',\n",
       " 'reopen',\n",
       " 'purchase',\n",
       " 'signup',\n",
       " 'purchase',\n",
       " 'reopen',\n",
       " 'signup']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "# (pathways2genes.loc[pathways2genes.hgnc_symbol_ids.apply(\n",
    "#     lambda x: shared_tfs(x)).sort_values(ascending=False)\n",
    "#     # .replace(0, np.nan)\n",
    "#     .dropna().index]\n",
    "#     .join(kegg_pathways, lsuffix='pathway', rsuffix='local', how='inner')\n",
    "# )[['pathway', 'global', 'local']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_pathways['mid_id'] += 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_pathways['local_id'] += 44"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>global</th>\n",
       "      <th>mid</th>\n",
       "      <th>hsaid</th>\n",
       "      <th>local</th>\n",
       "      <th>mid_id</th>\n",
       "      <th>global_id</th>\n",
       "      <th>local_id</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>150</td>\n",
       "      <td>Cellular Processes</td>\n",
       "      <td>Cell growth and death</td>\n",
       "      <td>4110</td>\n",
       "      <td>Cell cycle</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>176</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>151</td>\n",
       "      <td>Cellular Processes</td>\n",
       "      <td>Cell growth and death</td>\n",
       "      <td>4114</td>\n",
       "      <td>Oocyte meiosis</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>177</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>156</td>\n",
       "      <td>Cellular Processes</td>\n",
       "      <td>Cell growth and death</td>\n",
       "      <td>4115</td>\n",
       "      <td>p53 signaling pathway</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>178</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>152</td>\n",
       "      <td>Cellular Processes</td>\n",
       "      <td>Cell growth and death</td>\n",
       "      <td>4210</td>\n",
       "      <td>Apoptosis</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>193</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>153</td>\n",
       "      <td>Cellular Processes</td>\n",
       "      <td>Cell growth and death</td>\n",
       "      <td>4215</td>\n",
       "      <td>Apoptosis - multiple species</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>196</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>229</td>\n",
       "      <td>Organismal Systems</td>\n",
       "      <td>Nervous system</td>\n",
       "      <td>4730</td>\n",
       "      <td>Long-term depression</td>\n",
       "      <td>38</td>\n",
       "      <td>5</td>\n",
       "      <td>254</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>234</td>\n",
       "      <td>Organismal Systems</td>\n",
       "      <td>Sensory system</td>\n",
       "      <td>4740</td>\n",
       "      <td>Olfactory transduction</td>\n",
       "      <td>42</td>\n",
       "      <td>5</td>\n",
       "      <td>255</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>235</td>\n",
       "      <td>Organismal Systems</td>\n",
       "      <td>Sensory system</td>\n",
       "      <td>4742</td>\n",
       "      <td>Taste transduction</td>\n",
       "      <td>42</td>\n",
       "      <td>5</td>\n",
       "      <td>256</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328</th>\n",
       "      <td>233</td>\n",
       "      <td>Organismal Systems</td>\n",
       "      <td>Sensory system</td>\n",
       "      <td>4744</td>\n",
       "      <td>Phototransduction</td>\n",
       "      <td>42</td>\n",
       "      <td>5</td>\n",
       "      <td>257</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>236</td>\n",
       "      <td>Organismal Systems</td>\n",
       "      <td>Sensory system</td>\n",
       "      <td>4750</td>\n",
       "      <td>Inflammatory mediator regulation of TRP channels</td>\n",
       "      <td>42</td>\n",
       "      <td>5</td>\n",
       "      <td>258</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>330 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0              global                    mid  hsaid  \\\n",
       "0           150  Cellular Processes  Cell growth and death   4110   \n",
       "1           151  Cellular Processes  Cell growth and death   4114   \n",
       "2           156  Cellular Processes  Cell growth and death   4115   \n",
       "3           152  Cellular Processes  Cell growth and death   4210   \n",
       "4           153  Cellular Processes  Cell growth and death   4215   \n",
       "..          ...                 ...                    ...    ...   \n",
       "325         229  Organismal Systems         Nervous system   4730   \n",
       "326         234  Organismal Systems         Sensory system   4740   \n",
       "327         235  Organismal Systems         Sensory system   4742   \n",
       "328         233  Organismal Systems         Sensory system   4744   \n",
       "329         236  Organismal Systems         Sensory system   4750   \n",
       "\n",
       "                                                local  mid_id  global_id  \\\n",
       "0                                          Cell cycle      13          0   \n",
       "1                                      Oocyte meiosis      13          0   \n",
       "2                               p53 signaling pathway      13          0   \n",
       "3                                           Apoptosis      13          0   \n",
       "4                        Apoptosis - multiple species      13          0   \n",
       "..                                                ...     ...        ...   \n",
       "325                              Long-term depression      38          5   \n",
       "326                            Olfactory transduction      42          5   \n",
       "327                                Taste transduction      42          5   \n",
       "328                                 Phototransduction      42          5   \n",
       "329  Inflammatory mediator regulation of TRP channels      42          5   \n",
       "\n",
       "     local_id  count  \n",
       "0         176      1  \n",
       "1         177      1  \n",
       "2         178      1  \n",
       "3         193      1  \n",
       "4         196      1  \n",
       "..        ...    ...  \n",
       "325       254      1  \n",
       "326       255      1  \n",
       "327       256      1  \n",
       "328       257      1  \n",
       "329       258      1  \n",
       "\n",
       "[330 rows x 9 columns]"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_kegg = kegg_pathways.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_pathways['count'] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [],
   "source": [
    "sources2 = []\n",
    "targets2 = []\n",
    "\n",
    "# kegg_pathways = full_kegg.join(pathways2genes, lsuffix='local', rsuffix='pathway', how='inner').dropna()\n",
    "kegg_pathways = full_kegg\n",
    "\n",
    "labels2 = kegg_pathways[['global', 'global_id']].drop_duplicates().sort_values(by='global_id')['global'].values.tolist() + \\\n",
    "        kegg_pathways[['mid', 'mid_id']].drop_duplicates().sort_values(by='mid_id')['mid'].values.tolist() + \\\n",
    "        kegg_pathways[['local', 'local_id']].drop_duplicates().sort_values(by='local_id')['local'].values.tolist()\n",
    "\n",
    "values2 = []\n",
    "\n",
    "for l, global_id, mid_id, count in kegg_pathways[['global', 'global_id', 'mid_id', 'count']].groupby(['global', 'global_id', 'mid_id'])['count'].count().reset_index().values:\n",
    "    sources2.append(global_id)\n",
    "    targets2.append(mid_id)\n",
    "    values2.append(count)\n",
    "    \n",
    "    \n",
    "for l, global_id, mid_id, count in kegg_pathways[['mid', 'mid_id', 'local_id', 'count']].groupby(['mid', 'mid_id', 'local_id'])['count'].count().reset_index().values:\n",
    "    sources2.append(global_id)\n",
    "    targets2.append(mid_id)\n",
    "    values2.append(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fig = go.Figure(data=[go.Sankey(\n",
    "#     node=dict(\n",
    "#         thickness=20,  # default is 20\n",
    "#         line=dict(color=\"black\", width=0.9),\n",
    "#         label=labels2\n",
    "#     ),\n",
    "#     link=dict(\n",
    "#         source=sources2,\n",
    "#         target=targets2,\n",
    "#         value=values2,\n",
    "#         # label=time_to_next[:len(sources2)],\n",
    "#         # hovertemplate='%{value} unique users went from %{source.label} to %{target.label}.<br />' +\n",
    "#         # '<br />It took them %{label} in average.<extra></extra>',\n",
    "#     ))])\n",
    "\n",
    "# fig.update_layout(autosize=False, width=1400, height=6000, title_text=\"Pathways\",\n",
    "#                   font=dict(size=15), plot_bgcolor='white')\n",
    "\n",
    "\n",
    "# fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>global</th>\n",
       "      <th>mid</th>\n",
       "      <th>local</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Cellular Processes</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">Cell growth and death</th>\n",
       "      <th>Apoptosis</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Apoptosis - multiple species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell cycle</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cellular senescence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ferroptosis</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Organismal Systems</th>\n",
       "      <th>Nervous system</th>\n",
       "      <th>Synaptic vesicle cycle</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">Sensory system</th>\n",
       "      <th>Inflammatory mediator regulation of TRP channels</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olfactory transduction</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Phototransduction</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Taste transduction</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>330 rows × 0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [(Cellular Processes, Cell growth and death, Apoptosis), (Cellular Processes, Cell growth and death, Apoptosis - multiple species), (Cellular Processes, Cell growth and death, Cell cycle), (Cellular Processes, Cell growth and death, Cellular senescence), (Cellular Processes, Cell growth and death, Ferroptosis), (Cellular Processes, Cell growth and death, Necroptosis), (Cellular Processes, Cell growth and death, Oocyte meiosis), (Cellular Processes, Cell growth and death, p53 signaling pathway), (Cellular Processes, Cell motility, Regulation of actin cytoskeleton), (Cellular Processes, Cellular community - eukaryotes, Adherens junction), (Cellular Processes, Cellular community - eukaryotes, Focal adhesion), (Cellular Processes, Cellular community - eukaryotes, Gap junction), (Cellular Processes, Cellular community - eukaryotes, Signaling pathways regulating pluripotency of stem cells), (Cellular Processes, Cellular community - eukaryotes, Tight junction), (Cellular Processes, Transport and catabolism, Autophagy - animal), (Cellular Processes, Transport and catabolism, Autophagy - other), (Cellular Processes, Transport and catabolism, Endocytosis), (Cellular Processes, Transport and catabolism, Lysosome), (Cellular Processes, Transport and catabolism, Mitophagy - animal), (Cellular Processes, Transport and catabolism, Peroxisome), (Cellular Processes, Transport and catabolism, Phagosome), (Environmental Information Processing, Membrane transport, ABC transporters), (Environmental Information Processing, Signal transduction, AMPK signaling pathway), (Environmental Information Processing, Signal transduction, Apelin signaling pathway), (Environmental Information Processing, Signal transduction, Calcium signaling pathway), (Environmental Information Processing, Signal transduction, ErbB signaling pathway), (Environmental Information Processing, Signal transduction, FoxO signaling pathway), (Environmental Information Processing, Signal transduction, HIF-1 signaling pathway), (Environmental Information Processing, Signal transduction, Hedgehog signaling pathway), (Environmental Information Processing, Signal transduction, Hippo signaling pathway), (Environmental Information Processing, Signal transduction, Hippo signaling pathway - multiple species), (Environmental Information Processing, Signal transduction, Jak-STAT signaling pathway), (Environmental Information Processing, Signal transduction, MAPK signaling pathway), (Environmental Information Processing, Signal transduction, NF-kappa B signaling pathway), (Environmental Information Processing, Signal transduction, Notch signaling pathway), (Environmental Information Processing, Signal transduction, PI3K-Akt signaling pathway), (Environmental Information Processing, Signal transduction, Phosphatidylinositol signaling system), (Environmental Information Processing, Signal transduction, Phospholipase D signaling pathway), (Environmental Information Processing, Signal transduction, Rap1 signaling pathway), (Environmental Information Processing, Signal transduction, Ras signaling pathway), (Environmental Information Processing, Signal transduction, Sphingolipid signaling pathway), (Environmental Information Processing, Signal transduction, TGF-beta signaling pathway), (Environmental Information Processing, Signal transduction, TNF signaling pathway), (Environmental Information Processing, Signal transduction, VEGF signaling pathway), (Environmental Information Processing, Signal transduction, Wnt signaling pathway), (Environmental Information Processing, Signal transduction, cAMP signaling pathway), (Environmental Information Processing, Signal transduction, cGMP-PKG signaling pathway), (Environmental Information Processing, Signal transduction, mTOR signaling pathway), (Environmental Information Processing, Signaling molecules and interaction, Cell adhesion molecules (CAMs)), (Environmental Information Processing, Signaling molecules and interaction, Cytokine-cytokine receptor interaction), (Environmental Information Processing, Signaling molecules and interaction, ECM-receptor interaction), (Environmental Information Processing, Signaling molecules and interaction, Neuroactive ligand-receptor interaction), (Genetic Information Processing, Folding, sorting and degradation, Proteasome), (Genetic Information Processing, Folding, sorting and degradation, Protein export), (Genetic Information Processing, Folding, sorting and degradation, Protein processing in endoplasmic reticulum), (Genetic Information Processing, Folding, sorting and degradation, RNA degradation), (Genetic Information Processing, Folding, sorting and degradation, SNARE interactions in vesicular transport), (Genetic Information Processing, Folding, sorting and degradation, Sulfur relay system), (Genetic Information Processing, Folding, sorting and degradation, Ubiquitin mediated proteolysis), (Genetic Information Processing, Replication and repair, Base excision repair), (Genetic Information Processing, Replication and repair, DNA replication), (Genetic Information Processing, Replication and repair, Fanconi anemia pathway), (Genetic Information Processing, Replication and repair, Homologous recombination), (Genetic Information Processing, Replication and repair, Mismatch repair), (Genetic Information Processing, Replication and repair, Non-homologous end-joining), (Genetic Information Processing, Replication and repair, Nucleotide excision repair), (Genetic Information Processing, Transcription, Basal transcription factors), (Genetic Information Processing, Transcription, RNA polymerase), (Genetic Information Processing, Transcription, Spliceosome), (Genetic Information Processing, Translation, Aminoacyl-tRNA biosynthesis), (Genetic Information Processing, Translation, RNA transport), (Genetic Information Processing, Translation, Ribosome), (Genetic Information Processing, Translation, Ribosome biogenesis in eukaryotes), (Genetic Information Processing, Translation, mRNA surveillance pathway), (Human Diseases, Cancers: Overview, Central carbon metabolism in cancer), (Human Diseases, Cancers: Overview, Chemical carcinogenesis), (Human Diseases, Cancers: Overview, Choline metabolism in cancer), (Human Diseases, Cancers: Overview, MicroRNAs in cancer), (Human Diseases, Cancers: Overview, Pathways in cancer), (Human Diseases, Cancers: Overview, Proteoglycans in cancer), (Human Diseases, Cancers: Overview, Transcriptional misregulation in cancer), (Human Diseases, Cancers: Overview, Viral carcinogenesis), (Human Diseases, Cancers: Specific types, Acute myeloid leukemia), (Human Diseases, Cancers: Specific types, Basal cell carcinoma), (Human Diseases, Cancers: Specific types, Bladder cancer), (Human Diseases, Cancers: Specific types, Breast cancer), (Human Diseases, Cancers: Specific types, Chronic myeloid leukemia), (Human Diseases, Cancers: Specific types, Colorectal cancer), (Human Diseases, Cancers: Specific types, Endometrial cancer), (Human Diseases, Cancers: Specific types, Gastric cancer), (Human Diseases, Cancers: Specific types, Glioma), (Human Diseases, Cancers: Specific types, Hepatocellular carcinoma), (Human Diseases, Cancers: Specific types, Melanoma), (Human Diseases, Cancers: Specific types, Non-small cell lung cancer), (Human Diseases, Cancers: Specific types, Pancreatic cancer), (Human Diseases, Cancers: Specific types, Prostate cancer), (Human Diseases, Cancers: Specific types, Renal cell carcinoma), (Human Diseases, Cancers: Specific types, Small cell lung cancer), (Human Diseases, Cancers: Specific types, Thyroid cancer), (Human Diseases, Cardiovascular diseases, Arrhythmogenic right ventricular cardiomyopathy (ARVC)), ...]\n",
       "\n",
       "[330 rows x 0 columns]"
      ]
     },
     "execution_count": 264,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_pathways[['global', 'mid', 'local']].groupby(['global', 'mid', 'local']).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((6, 44), (44, 330), (330, 320))"
      ]
     },
     "execution_count": 275,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "global2mid = pd.DataFrame(index=kegg_pathways['global'].unique(), columns=kegg_pathways['mid'].unique()).fillna(0)\n",
    "mid2local = pd.DataFrame(index=kegg_pathways['mid'].unique(), columns=kegg_pathways['local'].unique()).fillna(0)\n",
    "local2tf = pd.DataFrame(index=kegg_pathways['local'].unique(), columns=data_csv.tf).fillna(0)\n",
    "global2mid.shape, mid2local.shape, local2tf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {},
   "outputs": [],
   "source": [
    "for from_, to_ in kegg_pathways[['global', 'mid']].values:\n",
    "    global2mid.at[from_, to_] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {},
   "outputs": [],
   "source": [
    "for from_, to_ in kegg_pathways[['mid', 'local']].values:\n",
    "    mid2local.at[from_, to_] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {},
   "outputs": [],
   "source": [
    "for from_, to_ in kegg_pathways.join(pathways2genes, lsuffix='local', rsuffix='pathway', how='inner').dropna()[['local', 'tf']].values:\n",
    "    for tf in to_:\n",
    "        local2tf.at[from_, tf]+=1\n",
    "        "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.5 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5 (default, Sep  4 2020, 07:30:14) \n[GCC 7.3.0]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "ba717acba6bbc10ecee861af44b959e6410892cef75714266950d58c109ec6e4"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
